05-lwt-cli-app/lib/repo/bibliography.ml

```diff
@@ -0,0 +1,78 @@
+module Row = struct
+  type t = {
+    book_id : int;
+    title : string;
+    first_name : string;
+    middle_name : string option;
+    last_name : string;
+  }
+end
+
+module Q = struct
+  open Caqti_request.Infix
+
+  (*
+    Caqti infix operators
+
+    ->! decodes a single row
+    ->? decodes zero or one row
+    ->* decodes many rows
+    ->. expects no row
+  *)
+
+  let insert =
+    Caqti_type.(tup2 int int ->. unit)
+      {|
+       INSERT INTO bibliography (author_id, book_id) VALUES (?, ?)
+      |}
+
+  let x = Caqti_type.(unit ->* tup2 (tup2 string string) string)
+
+  (* Example showing how to go beyond tup4 (there is no tup5) *)
```

Comment:
Maybe we should have targeted the upcoming release?  I don't expect to make any significant changes to the API for version 2.

Reply:
Yeah I did see that you had something in the works.

I'll be updating the examples to use the new tup variants provided by release 2.0 :+1:

Reply:
I've updated the examples to use the new tuple syntax: nice improvement!

I've also added a "custom type" example but it's a WIP: I'm encountering a little snag, not sure if it's a bug in my code or if I've found a bug in the library code.

Reply:
Oh, can I reproduce it with the current state of the PR?

Reply:
Yes. Here is how you can reproduce this problem.

Step 0: clone the v1 branch (I'm currently at commit 96c36ae850630b54edacfae06cfba4b87b9ee6b7)

```
benjamin@X230 ~/code/github.com/paurkedal/caqti-study/05-lwt-cli-app
$ opam list | grep caqti
caqti                         2.0.1        Unified interface to relational database libraries
caqti-async                   2.0.1        Async support for Caqti
caqti-driver-postgresql       2.0.1        PostgreSQL driver for Caqti based on C bindings
caqti-driver-sqlite3          2.0.1        Sqlite3 driver for Caqti using C bindings
caqti-lwt                     2.0.1        Lwt support for Caqti
```

Step 1: reset the database if necessary

```bash
make db-stop;rm -r ./local_db/;make db-init && make db-start && make db-create
```

Step 2: confirm the CLI app tests pass
```
cd ./05-lwt-cli-app/
PGHOST=localhost PGDATABASE=caqti_study PGPORT=5433 dune runtest
```

Step 3: uncomment "last_name" so that the Caqti custom type handles one more field from the underlying record.

Here's the git diff to clarify:

```
diff --git a/05-lwt-cli-app/lib/repo/author.ml b/05-lwt-cli-app/lib/repo/author.ml
index c5cdbc3..d466e4f 100644
--- a/05-lwt-cli-app/lib/repo/author.ml
+++ b/05-lwt-cli-app/lib/repo/author.ml
@@ -2,7 +2,7 @@ type t = {
   id : int;
   first_name : string;
       (* middle_name : string option; *)
-      (* last_name : string; *)
+  last_name : string;
 }
 
 module Q = struct
@@ -21,12 +21,12 @@ module Q = struct
     let intro id
         first_name
           (* middle_name  *)
-          (* last_name *) =
+        last_name =
       {
         id;
-        first_name
+        first_name;
         (* middle_name*)
-        (* last_name *);
+        last_name
       }
     in
     Caqti_type.Std.(
@@ -34,7 +34,7 @@ module Q = struct
       @@ proj int (fun x -> x.id)
       @@ proj string (fun x -> x.first_name)
       (* @@ proj (option string) (fun x -> x.middle_name) *)
-      (* @@ proj string (fun x -> x.last_name) *)
+      @@ proj string (fun x -> x.last_name)
       @@ proj_end)
 
   let insert =

```

Step 4a: the test "seed the database, then read many rows (alt, simple)" then fails with what looks like a decoding problem:

```
$ PGHOST=localhost PGDATABASE=caqti_study PGPORT=5433 dune runtest
Entering directory '/home/benjamin/code/github.com/paurkedal/caqti-study'
File "05-lwt-cli-app/test/repo/bibliography_test.ml", line 33, characters 0-532: seed the database, then read many rows (alt, simple) threw Postgresql.Error(_).
  Raised at Postgresql.result.check_field in file "src/postgresql.ml", line 646, characters 6-57
  Called from Postgresql.result#getvalue in file "src/postgresql.ml", line 697, characters 4-21
  Called from Caqti_driver_postgresql.decode_row.read_value in file "caqti-driver-postgresql/lib/caqti_driver_postgresql.ml", line 297, characters 33-52
  Called from Caqti_platform__Request_utils.decode_row.(fun) in file "caqti/lib-platform/request_utils.ml", line 174, characters 22-35
  Called from Caqti_driver_postgresql.decode_row.(fun) in file "caqti-driver-postgresql/lib/caqti_driver_postgresql.ml", line 307, characters 11-30
  Called from Caqti_driver_postgresql.Connect_functor.Make_connection_base.Response.fold.loop in file "caqti-driver-postgresql/lib/caqti_driver_postgresql.ml", line 635, characters 21-39
  Called from Caqti_driver_postgresql.Connect_functor.Make_connection_base.Response.fold.(fun) in file "caqti-driver-postgresql/lib/caqti_driver_postgresql.ml", line 639, characters 36-48
  Called from Caqti_platform__Connection_utils.Make_convenience.collect_list.f in file "caqti/lib-platform/connection_utils.ml", line 54, characters 17-48
  Called from Lwt.Sequential_composition.bind.create_result_promise_and_callback_if_deferred.callback in file "src/core/lwt.ml", line 1844, characters 16-19
  Re-raised at Lwt.Miscellaneous.poll in file "src/core/lwt.ml", line 3123, characters 20-29
  Called from Lwt_main.run.run_loop in file "src/unix/lwt_main.ml", line 27, characters 10-20
  Called from Lwt_main.run in file "src/unix/lwt_main.ml", line 106, characters 8-13
  Re-raised at Lwt_main.run in file "src/unix/lwt_main.ml", line 112, characters 4-13
  Called from Repo_test__Bibliography_test.(fun) in file "05-lwt-cli-app/test/repo/bibliography_test.ml", line 46, characters 4-33

FAILED 1 / 12 tests
```

Step 4b: I encounter a different error when running the query via the REPL:

```bash
PGHOST=localhost PGDATABASE=caqti_study PGPORT=5433 dune utop
```

```
─( 13:16:47 )─< command 0 >───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────{ counter: 0 }─
utop # open Repo;;
─( 13:16:48 )─< command 1 >───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────{ counter: 0 }─
utop # let conn = Init.connect_exn ();;
val conn : (module Caqti_lwt.CONNECTION) = <module>
─( 13:17:37 )─< command 2 >───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────{ counter: 0 }─
utop # Author.ls' conn;;
Exception: Postgresql.Error (Postgresql.Field_out_of_range (2, 2)).
Called from unknown location
Called from unknown location
Called from unknown location
Called from unknown location
Called from unknown location
Called from unknown location
Called from unknown location
Called from unknown location
Called from unknown location
Called from unknown location
Called from unknown location
─( 13:17:46 )─< command 3 >───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────{ counter: 0 }─
utop # 
```

Reply:
I think you forgot to adjust the query string for `ls'` to match the type.

Reply:
I had a second look to understand why the error messages are different.  I don't know why the location information is missing in utop, but the reason for the missing details about the exception in the unit test is that the printer for the exception has not been registered.  The following will reveal that it is the same error:
```diff
diff --git a/05-lwt-cli-app/test/repo/bibliography_test.ml b/05-lwt-cli-app/test/repo/bibliography_test.ml
index d725c11..90924c0 100644
--- a/05-lwt-cli-app/test/repo/bibliography_test.ml
+++ b/05-lwt-cli-app/test/repo/bibliography_test.ml
@@ -2,6 +2,11 @@ module Bibliography = Repo.Bibliography
 
 let str_error = Utils.str_error
 
+let () = Printexc.register_printer
+  (function
+   | Postgresql.Error err -> Some (Postgresql.string_of_error err)
+   | _ -> None)
+
 let%test_unit "seed the database, then read many rows" =
   let ( => ) =
     [%test_eq:
```
I'm not suggesting we add this there though.

Reply:
> I think you forgot to adjust the query string for `ls'` to match the type.

That was it! Sorry for the trouble.